test/data-driven/text-fragments-data-driven-test.js

```diff
@@ -0,0 +1,239 @@
+        directive,
+        )['text'];
+
+    if (processedDirectives.count == 0) {
+      return false;
+    }
+    const marks = processedDirectives[0];
+    const highlightingTextMatch =
+        marksArrayToString(marks) == testInput.highlightText;
+
+    return highlightingTextMatch;
+  }
+
+  /**
+   * Generates the output of a test case.
+   * @param {Boolean} generationSuccess - True if fragment generation test was
+   *     successful.
+   * @param {Boolean} highlightingSuccess - True if highlighting test was
+   *     successful.
+   * @return {String} The test's output text that will be compared with the
+   *     expected output in the test's output file.
+   */
+  generateOutput(generationSuccess, highlightingSuccess) {
+    return `GENERATION: ${this.getTextForTestResult(generationSuccess)}\n` +
+        `HIGHLIGHTING: ${this.getTextForTestResult(highlightingSuccess)}\n`
+  }
+
+  /**
+   * Helper method to get the textual representation of a test result that is
+   * included in the data driven test output.
+   * @param {Boolean} testSuccess - True if test was successful.
+   * @return {String} {@link OUTPUT_SUCCESS} when |testSuccess| is true
+   *     or {@link OUTPUT_FAILURE} when false.
+   */
+  getTextForTestResult(testSuccess) {
+    return testSuccess ? OUTPUT_SUCCESS : OUTPUT_FAILURE;
+  }
+
+  /**
+   * Text Fragments Data Driven Test input object.
+   * @typedef {Object} TextFragmentDataDrivenTestInput
+   * @property {String} htmlFileName - Html file name for the test case. This file should be in |htmlDirectory|.
+   * @property {String} [startParentId] - Id of the parent of the node where the
+   * test case’s selection starts. Optional: when null then the parent node is
+   * the document's body.
+   * @property {Number} startOffsetInParent - Offset of the selection’s starting
+   * node in its parent node.
+   * @property {Number} [startTextOffset] - When the selection starts in a text
+   * node, this value indicates the offset inside the node’s text where the
+   * selection starts.
+   * @property {String} [endParentId] - Id of the parent of the node where the
+   * test case’s selection ends. Optional: when null then the parent node is the
+   * document's body.
+   * @property {Number} endOffsetInParent - Offset of the selection’s ending
+   * node in its parent node.
+   * @property {Number} [endTextOffset] - When the selection ends in a text
+   * node, this value indicates the offset inside the node’s text where the
+   * selection ends.
+   * @property {String} selectedText - Expected text content inside the
+   * selection.
+   * @property {String} highlightText - Expected text content inside the range
+   * highlighted.
+   */
+
+  /**
+   * Parses the content of a test case input file.
+   * @param {String} input - Content of a test case input file.
+   * @return {TextFragmentDataDrivenTestInput} Test Case input object.
+   */
+  parseInput(input) {
```

Comment:
Should this be in the base class? Seems like all data driven tests will need to parse test input files

Reply:
Do you mean something like having an "abstract" implementation of parseInput in the base and make the subclass provide a concrete implementation?
Something like:

```
class Base {

  runDataDrivenTest(inputFile) {
    // read file content
    const input = this.parseInput(...);
    const testResult = this.generateResults(input);
   ...
  }
  
  generateResults(input) {
     // abstract: testing logic to be implemented in subclass
  }

  parseInput(inputText) {
    // abstract: input parsing to be implemented in subclass
  }
}

class Subclass extends Base {

  generateResults(input) {
     // concrete implementation of testing logic and output generation
  }

  parseInput(inputText) {
    // concrete implementation of input parsing e.g JSON.parse(inputText)
  }
}

Reply:
No I was suggesting to literally put

```
  /**
   * Parses the content of a test case input file.
   * @param {String} input - Content of a test case input file.
   * @return {TextFragmentDataDrivenTestInput} Test Case input object.
   */
  parseInput(input) {
    try {
      return JSON.parse(input);
    } catch (error) {
      throw new Error(
          'Invalid input file. Content must be a valid json object.\n' +
          `${error}.`);
    }
 }
```

in the base class, and the subclasses can choose to call that whenever it makes sense for them. Seems like we'll always be parsing an input file that we'll need to wrap in try/catch etc etc, so might as well avoid duplication and put it in the base class. If we need to parse non-JSON files, we could always have `parseJson()`, `parseString()` etc in the base class.

Just a suggestion, I don't really feel strongly either way!

Reply:
I understand. I think I feel the same way :). I do like the idea of moving it to base just to remove boilerplate code from the concrete test class implementation, leaving just code specific to the actual text fragment tests.

I'll do the refactoring. Thanks for your input!

Reply:
After giving it a second thought I also like having parseInput returning a specific object schema (TextFragmentDataDrivenTestInput), which I would lose if I move it to base. I'll go with parseInput on the subclass then and we can always refactor it if/when we have more implementations of data-driven tests.

Thanks again for reviewing and giving me useful suggestions.